File: chessmenthol/vision/tracker.py
from __future__ import annotations

from typing import Optional, Union

import chess
import numpy as np

from chessmenthol.position import (
    AssembledPosition,
    assemble,
    guess_orientation,
    guess_side_to_move,
)

from .capture import Capturer
from .detect import crop_squares, detect
from .pieces import PieceClassifier
from .types import Frame, square_name

ImageLike = Union[Frame, np.ndarray]


class Tracker:
    """Turns one captured frame into an AssembledPosition.

    Pipeline: grab -> detect -> crop_squares -> classify -> bridge to an 8x8
    geometric grid -> assemble. Orientation/side overrides let the UI correct
    the detection; they persist across frames (via prev_board for move inference).
    """

    def __init__(
        self,
        capturer: Optional[Capturer] = None,
        classifier: Optional[PieceClassifier] = None,
    ) -> None:
        self._capturer = capturer if capturer is not None else Capturer()
        self._classifier = classifier if classifier is not None else PieceClassifier()
        self._prev_board: Optional[chess.Board] = None
        self._orientation_override: Optional[str] = None
        self._side_override: Optional[chess.Color] = None

    def set_orientation_override(self, orientation: Optional[str]) -> None:
        self._orientation_override = orientation

    def set_side_override(self, side: Optional[chess.Color]) -> None:
        self._side_override = side

    def set_region(self, region) -> None:
        self._capturer.set_region(region)

    def grab_full_desktop(self):
        return self._capturer.grab_full_desktop()

    def reset(self) -> None:
        self._prev_board = None

    def detect_position(
        self, frame: Optional[ImageLike] = None
    ) -> Optional[AssembledPosition]:
        if frame is None:
            frame = self._capturer.grab()
        location = detect(frame)
        if location is None:
            return None
        crops = crop_squares(frame, location)
        labels = self._classifier.classify(crops)

        # Bridge: recover the geometric grid using the SAME orientation crop_squares
        # named the crops with (location.orientation_hint). assemble then applies the
        # resolved orientation, so an override flips the chess mapping without re-cropping.
        label_by_name = {c.square: lab for c, lab in zip(crops, labels)}
        grid = [
            [label_by_name[square_name(col, row, location.orientation_hint)] for col in range(8)]
            for row in range(8)
        ]

        orientation = (
            self._orientation_override
            or location.orientation_hint
            or guess_orientation(grid)
            or "white_bottom"
        )
        side = self._resolve_side(grid, orientation, location)
        assembled = assemble(
            grid, orientation=orientation, side_to_move=side, prev_board=self._prev_board
        )
        if assembled.is_legal:
            self._prev_board = assembled.board
        return assembled

    def _resolve_side(self, grid, orientation, location) -> chess.Color:
        if self._side_override is not None:
            return self._side_override
        # Two-pass: assemble provisionally to get a board for the highlight/move guess.
        provisional = assemble(
            grid, orientation=orientation, side_to_move=chess.WHITE, prev_board=self._prev_board
        )
        if provisional.board is None:
            return chess.WHITE
        return guess_side_to_move(
            provisional.board,
            prev_board=self._prev_board,
            move=provisional.move,
            highlight_squares=location.highlight_squares,
        )

File: frontend/src/vision/tracker.ts
/**
 * vision/tracker.ts — single-frame pipeline (classifier-injected).
 *
 * Faithful TS port of the original Python chessmenthol/vision/tracker.py (removed in the Svelte+Tauri migration) MINUS the capturer.
 * Capture lives in `lib/capture.ts`; this module only receives an RgbaImage.
 *
 * Pipeline: detectPosition(image)
 *   detect(image)
 *   → cropSquares(image, location)
 *   → classifier.classify(crops)              [async — ONNX in prod]
 *   → bridge: recover the geometric grid with location.orientationHint
 *   → resolve orientation (override ?? hint ?? guess ?? 'white_bottom')
 *   → resolve side (_resolveSide: two-pass provisional assemble + guessSideToMove)
 *   → assemble(grid, { orientation, white: side, prevFen })
 *   → update prevFen when legal; return AssembledPosition.
 *
 * prevFen is a STRING (the assembled.fen from the previous legal frame).
 * Move inference happens inside assemble via inferMove — we just thread prevFen.
 */

import { detect, cropSquares } from './detect';
import { assemble, guessOrientation, guessSideToMove } from './position';
import type { AssembledPosition, SquareLabel } from './position';
import { squareName } from './types';
import type { Orientation, RgbaImage, SquareImage, BoardLocation } from './types';

// ─── ClassifierLike ──────────────────────────────────────────────────────────

/** Minimal interface for an async piece classifier. Satisfied by PieceClassifier
 *  (onnxruntime-backed) and the FakeClassifier used in tests. */
export interface ClassifierLike {
  classify(crops: SquareImage[]): Promise<SquareLabel[]>;
}

// ─── Tracker ─────────────────────────────────────────────────────────────────

export class Tracker {
  /** FEN string of the previous legal frame, or null if no frame has been seen. */
  private prevFen: string | null = null;

  /** null = auto-detect; true = white; false = black. */
  private sideOverride: boolean | null = null;

  /** null = use hint / guess. */
  private orientationOverride: Orientation | null = null;

  constructor(private readonly classifier: ClassifierLike) {}

  setSideOverride(white: boolean | null): void {
    this.sideOverride = white;
  }

  setOrientationOverride(o: Orientation | null): void {
    this.orientationOverride = o;
  }

  /** Clear prevFen so the next frame cannot infer a move from this frame. */
  reset(): void {
    this.prevFen = null;
  }

  /**
   * Detect the board in `image` and return the assembled position, or null if
   * no board is found. Updates prevFen when the assembled position is legal.
   */
  async detectPosition(image: RgbaImage): Promise<AssembledPosition | null> {
    const location = detect(image);
    if (location === null) return null;

    const crops = cropSquares(image, location);
    const labels = await this.classifier.classify(crops);

    // Bridge: recover the geometric grid using the SAME orientation hint that
    // cropSquares used to name the crops. An override applied later flips the
    // chess mapping without re-cropping.
    const labelByName = new Map<string, SquareLabel>();
    for (let i = 0; i < crops.length; i++) {
      labelByName.set(crops[i].square, labels[i]);
    }

    const grid: SquareLabel[][] = [];
    for (let row = 0; row < 8; row++) {
      const gridRow: SquareLabel[] = [];
      for (let col = 0; col < 8; col++) {
        const name = squareName(col, row, location.orientationHint);
        // Fall back to empty+zero-confidence if a square is somehow missing.
        gridRow.push(labelByName.get(name) ?? { piece: null, confidence: 0 });
      }
      grid.push(gridRow);
    }

    const orientation: Orientation =
      this.orientationOverride ??
      location.orientationHint ??
      guessOrientation(grid) ??
      'white_bottom';

    const white = this._resolveSide(grid, orientation, location);
    const assembled = assemble(grid, { orientation, white, prevFen: this.prevFen });

    // Advance prevFen only on legal frames: an illegal detection (noise /
    // misclassification) must not corrupt the cross-frame move-inference baseline.
    if (assembled.isLegal) {
      this.prevFen = assembled.fen;
    }

    return assembled;
  }

  // ─── _resolveSide ──────────────────────────────────────────────────────────

  /**
   * Mirror of Python's _resolve_side.
   *
   * If there is a user override, use it directly.
   * Otherwise do a two-pass provisional assemble (white-to-move) to get a
   * legal board, then ask guessSideToMove. Falls back to white when the
   * provisional board is illegal (no board).
   *
   * Two passes are needed because `assemble` requires a `white` side-to-move
   * argument to build a legal board, but the side cannot be known without a
   * legal board first; the provisional pass assumes white to obtain one, then
   * guessSideToMove corrects it.
   *
   * Returns true = white, false = black.
   */
  private _resolveSide(
    grid: SquareLabel[][],
    orientation: Orientation,
    location: BoardLocation,
  ): boolean {
    if (this.sideOverride !== null) {
      return this.sideOverride;
    }

    const provisional = assemble(grid, {
      orientation,
      white: true,
      prevFen: this.prevFen,
    });

    if (!provisional.isLegal) {
      return true; // default: white
    }

    const side = guessSideToMove(provisional.fen, {
      prevFen: this.prevFen,
      move: provisional.move,
      highlightSquares: location.highlightSquares,
    });

    return side === 'white';
  }
}
